[Segmentation] Added generalized dice score metric

jlcsilva opened the pull request:
## What does this PR do?

Fixes #1089 . Didn't manage to make it work for multidimensional multilabel scenarios. However, that shouldn't a very common use case.

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

Borda commented:
I beilve that this shall be automatically resolved in cmd by `git merge upstream/master`

Borda commented:
@jlcsilva shall be resolved now, pls check if I did not make a mistake... :rabbit:

jlcsilva commented:
> @jlcsilva shall be resolved now, pls check if I did not make a mistake... 🐰

Seems ok to me. 

Also, I know why the tests are failing, but I'm unsure about what to do. In the file tests/unittests/classification/inputs.py
I changed the data type of some input tensors, such as _input_binary, from int to float, to match the description [here](https://torchmetrics.readthedocs.io/en/latest/pages/classification.html#input-types). Since the tests are failing, I'm not sure what's wrong, the documentation or the code. Any thoughts from someone with a better overall view of the repo?

SkafteNicki commented:
Hi @jlcsilva, when looking at the changes in the PR there also seems to be some changes to other metrics. Any particular reason for this?

jlcsilva commented:
> Hi @jlcsilva, when looking at the changes in the PR there also seems to be some changes to other metrics. Any particular reason for this?

In the StatScore class I added the None option for the reduce argument, which I needed for my generalized dice score implementation, and changed its documentation and that of its subclasses accordingly. Additionally, I corrected some typos and added some missing information to these classes' documentation. 

Finally, I changes the data type of some test inputs to match that in the general input type documentation page, and added new multidimensional input tensors. To me, this seems to be what is making the code fail the tests. Currently, I'm out of office, but I'll check it as soon as I go back to work

Borda commented on src/torchmetrics/classification/generalized_dice.py:
```diff
@@ -0,0 +1,169 @@
+            Determines whether the input is multiclass (if True) or multilabel (if False). Defaults to True.
+
+        multidim:
+            Determines whether the input is multidim or not. Defaults to True.
```
```suggestion
            Determines whether the input is ``multidim`` or not.
```

Borda replied:
```suggestion
            Determines whether the input is multidim or not.
```

Borda commented on src/torchmetrics/classification/generalized_dice.py:
```diff
@@ -0,0 +1,169 @@
+
+        zero_division:
+            The value to use for the score if denominator equals zero. If set to None, the score will be 1 if the
+            numerator is also 0, and 0 otherwise. Defaults to None.
```
```suggestion
            numerator is also 0, and 0 otherwise.
```

Borda commented on src/torchmetrics/functional/classification/generalized_dice.py:
```diff
@@ -0,0 +1,247 @@
+        weights[..., ignore_index] = 0
+        tp[..., ignore_index] = 0
+        fp[..., ignore_index] = 0
+        fn[..., ignore_index]
```
this seems to be void

Borda commented on src/torchmetrics/functional/classification/generalized_dice.py:
```diff
@@ -0,0 +1,247 @@
+
+        zero_division:
+            The value to use for the score if denominator equals zero. If set to None, the score will be 1 if the
+            numerator is also 0, and 0 otherwise. Defaults to None.
```
```suggestion
            numerator is also 0, and 0 otherwise.
```

Borda commented on src/torchmetrics/functional/classification/generalized_dice.py:
```diff
@@ -0,0 +1,247 @@
+        num_classes:
+            Number of classes.
+
+        multiclass:
+            Determines whether the input is multiclass (if True) or multilabel (if False). Defaults to True.
+
+        multidim:
+            Determines whether the input is multidim or not. Defaults to True.
```
```suggestion
        multiclass:
            Determines whether the input is multiclass (if True) or multilabel (if False).

        multidim:
            Determines whether the input is multidim or not.
```
duplicated from func signature

Borda commented on src/torchmetrics/classification/generalized_dice.py:
```diff
@@ -0,0 +1,169 @@
+            The default value (``None``) will be interpreted as 1.
+
+        multiclass:
+            Determines whether the input is multiclass (if True) or multilabel (if False). Defaults to True.
```
```suggestion
            Determines whether the input is multiclass (if True) or multilabel (if False).
```

Borda commented on tests/classification/test_generalized_dice_score.py:
```diff
@@ -0,0 +1,307 @@
+seed_all(42)
+
+
+def _sk_generalized_dice(
```
I do not see any import or use of scikit-learn so why do you name it this way?

SkafteNicki reviewed (changes requested):
Hi @jlcsilva,
Sorry for the long delay. We are currently doing a major refactor of the classification package #1001 and are therefore putting this PR on hold for now. When the refactor is done we will mix merge conflicts and afterwards focus on getting your awesome work merged :)

jlcsilva commented:
Hey @SkafteNicki,
No problem! I've also been away, so wasn't able to help. I noticed the changes when I came back and pulled the repo again. I'm actually a bit busy today and next week, but let me know if there is something I can help with :)

Borda commented:
@jlcsilva @SkafteNicki, mind checking the collisions after the refactor was merged... :)

Borda commented:
@jlcsilva @stancld, mind checking the collisions after the refactor was merged... :otter:

stancld commented:
@Borda I'll do rebase now.

stancld commented:
ℹ️ It will be needed to add test cases after refactoring tests as a part of the classification refactor.

japui-coder reviewed (changes requested):
tests/integrations/test_lightning.py

SkafteNicki commented:
Going to take over this PR now so we hopefully can land it within the next week :]

Borda commented:
> Going to take over this PR now so we hopefully can land it within the next week :]

hi @SkafteNicki how is it going here? :)

Borda commented:
@justusschock, how much need to be done? :chipmunk:

Borda commented on src/torchmetrics/classification/dice.py:
```diff
@@ -18,14 +18,18 @@
 class Dice(Metric):
     r"""Compute `Dice`_.
 
+    .. deprecated:: v0.10
```
```suggestion
    .. deprecated:: v0.12
```

Borda commented on src/torchmetrics/classification/dice.py:
```diff
@@ -18,14 +18,18 @@
     r"""Compute `Dice`_.
 
+    .. deprecated:: v0.10
+        The `Dice` module was deprecated in v0.10 and will be removed in v0.11. Use `F1Score` module instead which
```
```suggestion
        The `Dice` module was deprecated in v0.12 and will be removed in v0.13. Use `F1Score` module instead which
```

Borda commented on src/torchmetrics/classification/dice.py:
```diff
@@ -145,7 +149,12 @@ def __init__(
     ) -> None:
-        super().__init__(**kwargs)
+        rank_zero_warn(
+            "The `dice` function was deprecated in v0.10 and will be removed in v0.11. Use `f1score` function instead"
```
```suggestion
            "The `dice` function was deprecated in v0.12 and will be removed in v0.13. Use `f1score` function instead"
```

Borda commented on src/torchmetrics/classification/generalized_dice.py:
```diff
@@ -0,0 +1,247 @@
+            If ``top_k`` is not an ``integer`` larger than ``0``.
+
+    Example:
+        >>> import torch
+        >>> from torchmetrics import GeneralizedDiceScore
+        >>> preds = torch.tensor([2, 0, 2, 1])
+        >>> target = torch.tensor([1, 1, 2, 0])
```
```suggestion
        >>> from torch import tensor
        >>> from torchmetrics import GeneralizedDiceScore
        >>> preds = tensor([2, 0, 2, 1])
        >>> target = tensor([1, 1, 2, 0])
```

Borda commented on src/torchmetrics/functional/classification/generalized_dice.py:
```diff
@@ -0,0 +1,413 @@
+        ignore_index: Optional index of the class to ignore in the score computation
+        zero_division: The value to use for the score if denominator equals zero. If set to 0, score will be 1
+            if the numerator is also 0, and 0 otherwise
+
+    Example:
+        >>> from torchmetrics.functional.classification.stat_scores import _stat_scores_update
+        >>> from torchmetrics.functional.classification.generalized_dice import _generalized_dice_compute
+        >>> preds  = torch.tensor([2, 0, 2, 1])
+        >>> target = torch.tensor([1, 1, 2, 0])
+        >>> tp, fp, tn, fn = _stat_scores_update(preds, target, reduce='macro', mdmc_reduce='samplewise')
+        >>> _generalized_dice_compute(tp, fp, fn, average='samples')
+        tensor(0.3478)
```
```suggestion
```
protected does not need examples

Borda commented on src/torchmetrics/functional/classification/generalized_dice.py:
```diff
@@ -0,0 +1,413 @@
+            If ``top_k`` is not an integer larger than ``0``.
+
+    Example:
+        >>> import torch
+        >>> from torchmetrics.functional import generalized_dice_score
+        >>> preds = torch.tensor([2, 0, 2, 1])
+        >>> target = torch.tensor([1, 1, 2, 0])
```
```suggestion
        >>> from torch import tensor
        >>> from torchmetrics.functional import generalized_dice_score
        >>> preds = tensor([2, 0, 2, 1])
        >>> target = tensor([1, 1, 2, 0])
```

Borda reviewed (approved):
except https://github.com/Lightning-AI/metrics/pull/1090#discussion_r909536311it looks good to me

Borda commented:
@jlcsilva, any chance you could work on this one, in particular, add some tests... :flags:

Borda commented:
seems we hit some internal Gh issue, will be restarted later...

Borda reviewed (approved):
So happy to see it landing 🤩
